### DISCUSSION TITLE
setEdges in onConnect doesn't update the edges variable

### DISCUSSION THREAD

**Discussion by Cashmeiser** [xyflow/xyflow#4038] [Q&A]
Inside   `const onConnect = useCallback(...` I do some checks on current edges do decide if it is an allowed connection. If it is the edge is added with:
`setEdges(els => addEdge(params, els));`

Visually I can see the edge being created. But the edge const seems not to be updated with the new value.
I initialize it with   `const [edges, setEdges, onEdgesChange] = useEdgesState([]);`

In the onConnect I do some check that edge array doesn't contain an edge with some values making the connection  illegal. But if I add an edge that shouldn't be legal it work at first time when edges i loaded from backend. But when I add edges they aren't in the edges-const until I make a "save" and it will be updated.
Save looks something like:
```
  const onClickSaveStrategy = useCallback(() => {
    if (reactFlowInstance) {
      const flow = reactFlowInstance.toObject();
      const jsonFlow = JSON.stringify(flow);
      dispatch(postSaveStrategy({strategyName, flow: jsonFlow})).then(() => dispatch(getStoredUserStrategies()));
    }
  }, [reactFlowInstance, strategyName]);
```

What am I missing? How do I make the edge-array being updated directly when edges are added? Having same issue with Nodes...

**Comment by bcakmakoglu** [comment]
Can you share the `onConnect` part of your code? Might be there's something wrong with that.

**Reply by Cashmeiser** [reply]
Yes, here is the onConnect with comments.
I have a base node with two edges which have the trade type of either "enter" or "exit" (and nodes not connected to flow with base node have type as undefined) If a node is already connected with in the "enter"-flow it cannot get an edge connected with the tradeType=exit.
```
  const onConnect = useCallback(
    params => {
      const sourceNode = nodes.find(e => e.id === params.source);
      console.log(`onConnect sourceNode stringify = ${JSON.stringify(sourceNode, null, 2)}`);

      let tradeType: TradeType;

      // If sourceNode is of type baseDataNode or sourceNode have an edge of type entryOrExitEdge
      if (sourceNode && sourceNode.type === 'baseDataNode') {
        // first in flow, removed code since not  hit for this example
      } else {
        // Check if the source node has an edge connected with tradeType set to ENTRY or EXIT
        const upstreamConnectedEdge = edges.find(edge => edge.target === params.source && edge?.data?.tradeType);
        tradeType = upstreamConnectedEdge?.data?.tradeType; // ENTRY, EXIT or undefined
        params.type = 'coloredTradeTypeEdge';
      }

      params.data = {...params.data, tradeType: tradeType};

      // if the node already have edge with opposite tradeType it cannot be connected (cannot be in both ENTRY and EXIT flow)
      const isConnectedWithOppositeTradeType = edges.find(edge => edge.target === params.target && edge?.data?.tradeType && edge.data.tradeType !== tradeType);
      // Check whole downstream is not connected to different color... if the node (or any child-node) already have edge with opposite tradeType it cannot be connected (cannot be in both ENTRY and EXIT flow)
      if (!hasDownstreamFlowWithOppositeTradeType(params.target, tradeType)) {
        // Add the new edge:
        setEdges(els => addEdge(params, els));
        // Update edges downstream with the same tradeType (could been connected with other nodes but with tradeType=undefined)
        updateDownstreamEdges(params.target, tradeType);
      }
    },
    [nodes, setEdges, setNodes]
  );
```

In the whole `return` for the react function I have an outer `<ReactFlowProvider>` and inside I have the ReactFlow:

                    <div className="reactflow-wrapper card" ref={reactFlowWrapper}>
                      <ReactFlow
                        // elements={elements}
                        // onElementsRemove={onElementsRemove}
                        // onLoad={onLoad}
                        nodes={nodes}
                        edges={edges}
                        onNodesChange={onNodesChange}
                        onEdgesChange={onEdgesChange}
                        onEdgeUpdate={onEdgeUpdate}
                        onEdgeUpdateStart={onEdgeUpdateStart}
                        onEdgeUpdateEnd={onEdgeUpdateEnd}
                        onConnect={onConnect}
                        onInit={setReactFlowInstance}
                        onDrop={onDrop}
                        onDragOver={onDragOver}
                        nodeTypes={nodeTypes} // https://reactflow.dev/docs/guides/custom-nodes/#adding-the-node-type
                        edgeTypes={edgeTypes}
                        connectionRadius={100}
                        snapToGrid={true}
                        snapGrid={[25, 25]}
                      >
                        <Background/>
                        <Controls/>
                      </ReactFlow>
                    </div>

**Comment by Cashmeiser** [comment]
Update; I got it to work by updating `onConnect` to not have `useCallback`.
Maybe it can be but with dependent on something more than just:
`[nodes, setEdges, setNodes]`
